File: leetcode/556-next-greater-element-iii/main.go
package main

import (
	"fmt"
	"math"
)

/*
	classic approach: Next lexicographical permutation algorithm(use a stack)
	e.g. 43143221
	- from the end, stack up the item if num[i] > num[i+1]
	- once it encounters a smaller number from the end, this is the target we want
		e.g. 43 <1> 43221
	- i need to swap the target with the value in the stack which is just larger then it
		e.g. 43 <1> 43221 => 43 <2> 43211
	- reverse the left half and put it back to the number
		e.g. 43 <2> 43211 => 43 <2> 11234
	- combine them together and form the result
		e.g. 43 <2> 11234 => 43211234

	ref:
	- https://www.nayuki.io/page/next-lexicographical-permutation-algorithm

	Time	O(n)
	Space	O(n)
	0 ms, faster than 100.00%
*/
func nextGreaterElement(n int) int {
	// from the end, put the ascending items into a stack
	// and find the target for swapping
	remain := n
	stack := []int{}
	target := -1
	for remain > 0 {
		cur := remain % 10
		remain /= 10
		if len(stack) == 0 || cur >= stack[len(stack)-1] {
			stack = append(stack, cur)
		} else {
			target = cur
			break
		}
	}
	if target == -1 {
		return -1
	}
	// find the least larger number(>target) in the stack
	// and swap the target with the least larger number
	next := -1
	for i := 0; i < len(stack); i++ {
		if stack[i] > target {
			next = stack[i]
			stack[i] = target
			break
		}
	}
	// combine the stack, next number, and remain to form the result
	res := 0
	cnt := 0
	for len(stack) > 0 {
		pop0 := stack[0]
		stack = stack[1:]
		res = res*10 + pop0
		cnt++
	}
	res = res + next*int(math.Pow10(cnt))
	cnt++
	res = remain*int(math.Pow10(cnt)) + res
	if res > math.MaxInt32 {
		return -1
	}
	return res
}

func main() {
	fmt.Println(nextGreaterElement(21))
	fmt.Println(nextGreaterElement(12))
	fmt.Println(nextGreaterElement(43143221))
	fmt.Println(nextGreaterElement(1999999999))

	fmt.Println(nextGreaterElement(0))
	fmt.Println(nextGreaterElement(1))
}

File: leetcode/556-next-greater-element-iii/main.py
"""
    classic approach: Next lexicographical permutation algorithm(use a stack)

    similar to lc31

	e.g. 43143221
	- find the non-increasing suffix, e.g. 431<43221>
	- once it encounters a smaller number from the end, this is the target we want
		e.g. 43 <1> 43221
	- i need to swap the target with the value in the stack which is just larger then it
		e.g. 43 <1> 43221
                       ^
         => 43 <2> 43211
                      ^
	- reverse the right half and put it back to the number
		e.g. 43 <2> 43211 => 43 <2> 11234
	- combine them together and form the result
		e.g. 43 <2> 11234 => 43211234

	ref:
	- https://www.nayuki.io/page/next-lexicographical-permutation-algorithm

	Time	O(n)
	Space	O(n)
	8 ms, faster than 100.00%
"""


class Solution(object):
    def nextGreaterElement(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n < 0:
            return -1
        # transform the digits to an array
        nums = []
        while n > 0:
            nums = [n % 10] + nums
            n /= 10
        # find the pivot from the increasing sequence from the right
        pivot = -1
        for i in range(len(nums)-2, -1, -1):
            if nums[i] < nums[i+1]:
                pivot = i
                break
        # if no pivot, return fail to get the next one
        if pivot == -1:
            return -1
        # find the number to swap
        target = -1
        for i in range(len(nums)-1, -1, -1):
            if nums[i] > nums[pivot]:
                target = i
                break
        # swap
        nums[pivot], nums[target] = nums[target], nums[pivot]
        # reverse the increasing sequence from the right
        self.reverse(nums, pivot+1, len(nums)-1)
        # transform back to an integer
        res = 0
        for num in nums:
            res = res*10 + num
        # check range
        if res > 2**31-1:
            return -1
        return res

    def reverse(self, nums, start, end):
        left = start
        right = end
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1


print(Solution().nextGreaterElement(12))
print(Solution().nextGreaterElement(21))
print(Solution().nextGreaterElement(312))
print(Solution().nextGreaterElement(43143221))
print(Solution().nextGreaterElement(1999999999))
print(Solution().nextGreaterElement(0))
print(Solution().nextGreaterElement(1))
